src/xwayland/xserver.rs

```diff
@@ -0,0 +1,358 @@
+            eprintln!("[smithay] XWayland first fork failed: {:?}", e);
+            return Err(());
+        }
+    };
+
+    guard.instance = Some(XWaylandInstance {
+        display_lock: lock,
+        wayland_client: client,
+        sigusr1_handler: Some(sigusr1_handler),
+        wm_fd: Some(x_wm_me),
+        started_at: creation_time,
+        child_pid: Some(child_pid),
+    });
+
+    Ok(())
+}
+
+impl<WM: XWindowManager> Inner<WM> {
+    // Shutdown the XWayland server and cleanup everything
+    fn shutdown(&mut self) {
+        // don't do anything if not running
+        if let Some(mut instance) = self.instance.take() {
+            self.wm.xwayland_exited();
+            // kill the client
+            instance.wayland_client.kill();
+            // remove the event source
+            if let Some(s) = instance.sigusr1_handler.take() {
+                s.remove();
+            }
+            // All connexions and lockfiles are cleaned by their destructors
+
+            // Remove DISPLAY from the env
+            ::std::env::remove_var("DISPLAY");
+            // We do like wlroots:
+            // > We do not kill the Xwayland process, it dies to broken pipe
+            // > after we close our side of the wm/wl fds. This is more reliable
+            // > than trying to kill something that might no longer be Xwayland.
+        }
+    }
+}
+
+fn client_destroy<WM: XWindowManager + 'static>(data: *mut ()) {
+    let inner = unsafe { Rc::from_raw(data as *const () as *const RefCell<Inner<WM>>) };
+
+    // shutdown the server
+    let started_at = inner.borrow().instance.as_ref().map(|i| i.started_at);
+    inner.borrow_mut().shutdown();
+
+    // restart it, unless we really just started it, if it crashes right
+    // at startup there is no point
+    if started_at.map(|t| t.elapsed().as_secs()).unwrap_or(10) > 5 {
+        let _ = launch(&inner);
+    }
+}
+
+fn xwayland_ready<WM: XWindowManager>(inner: &Rc<RefCell<Inner<WM>>>) {
+    use nix::sys::wait;
+    let mut guard = inner.borrow_mut();
+    let inner = &mut *guard;
+    // instance should never be None at this point
+    let instance = inner.instance.as_mut().unwrap();
+    let wm = &mut inner.wm;
+    // neither the pid
+    let pid = instance.child_pid.unwrap();
+
+    // find out if the launch was a success by waiting on the intermediate child
+    let mut success: bool;
+    loop {
+        match wait::waitpid(pid, None) {
+            Ok(wait::WaitStatus::Exited(_, 0)) => {
+                // XWayland was correctly started :)
+                success = true;
+                break;
+            }
+            Err(NixError::Sys(Errno::EINTR)) => {
+                // interupted, retry
+                continue;
+            }
+            _ => {
+                // something went wrong :(
+                success = false;
+                break;
+            }
+        }
+    }
+
+    if success {
+        // signal the WM
+        wm.xwayland_ready(
```

Comment:
So if this does not succeed, how would a user of smithay know when to retry starting XWayland?

Reply:
Well, the reasoning if that if XWayland crashed at startup, there is no real point restarting it later, it's unlikely to change anything.

It may be worth logging, though. Event if the Xorg logs will already contain a lot of things probably.